packages/perseus/src/widgets/interactive-graphs/mafs-graph.test.tsx

```diff
@@ -339,4 +332,178 @@ describe("MafsGraph", () => {
         expect(line.getAttribute("x2")).toBe(expectedX2 + "");
         expect(line.getAttribute("y2")).toBe(-expectedY2 + "");
     });
+
+    it("MovableLine moves up based on up keystroke ", async () => {
```

Comment:
Added tests for movement in each direction. I know this can also be written as a single test that runs through an array of options and plugs those values in, but since it was only 4 I didn't bother. Also I decided to use different coordinates for the line each time, which wouldn't work with the single test approach. Totally not necessary to use different points too though.

Reply:
I'm not sure if these tests actually protect us against regressions. In any case, I think the developer experience of running them could be improved. When I tried bringing back the old logic in `movable-line.tsx`, I got a weird failure:

```
Error: expect(jest.fn()).toHaveBeenCalledWith(...expected)

Expected: {"delta": [0, 0.5], "itemIndex": 0, "type": "move-line"}
Received
       1
          Object {
        -   "delta": Array [
        -     0,
        -     0.5,
        +   "snapStep": Array [
        +     1,
        +     1,
            ],
        -   "itemIndex": 0,
        -   "type": "move-line",
        +   "type": "change-snap-step",
          },
       2
          Object {
        -   "delta": Array [
        -     0,
        -     0.5,
        +   "range": Array [
        +     Array [
        +       -10,
        +       10,
        +     ],
        +     Array [
        +       -10,
        +       10,
              ],
        -   "itemIndex": 0,
        -   "type": "move-line",
        +   ],
        +   "type": "change-range",
          },
       3
          Object {
            "delta": Array [
              0,
        -     0.5,
        +     0.4,
            ],
            "itemIndex": 0,
            "type": "move-line",
          },

Number of calls: 4
```

It looks like there are some other actions getting dispatched that we don't care about. I think the part we actually care about is this:

```
        -     0.5,
        +     0.4,
```

but even that doesn't tell the developer running the test "the line isn't going to move because Mafs isn't snapping the point correctly".

Looking at the failure, I also noticed that a `change-snap-step` action is getting dispatched which sets the snap step to `[1, 1]`. This will probably interfere with the test logic. I think that action is getting dispatched because `baseMafsGraphProps` has a snapStep of `[1, 1]`.

Long story short, I think that rather than mock `dispatch`, I'd prefer to simulate the `useReducer` store like this:

```ts
// at the top level:

function createFakeStore<S, A>(reducer: (state: S, action: A) => S, state: S) {
    return {
        dispatch(action: A) {
            state = reducer(state, action);
        },

        getState() {
            return state;
        }
    }
}

// ...and then, in each test:

  const {dispatch, getState} = createFakeStore(interactiveGraphReducer, initialState)

  const baseMafsGraphProps = getBaseMafsGraphProps();

  render(
      <MafsGraph
          state={getState()}
          dispatch={dispatch}
          {...baseMafsGraphProps}
          snapStep={[0.5, 0.5]}
      />,
  );

  const group = screen.getByTestId("movable-line");
  group.focus();
  await userEvent.keyboard("[ArrowDown]");

  expect(getState().coords).toEqual(expectedCoords);
```

If we do that, it's easy to see from the test failure where the line ended up (whether it moved or not). We don't have to convert the line coordinates to pixel coordinates or re-render the graph.

I'd be happy to pair on reworking the tests if you want. Maybe tomorrow?

Reply:
Awesome! Thanks for the updated test suggestion! Yea, I'll play around with this a bit and see what I can manage. Definitely pairing if I can't get it working. I just took over a test that was there (and no longer needed), so I definitely appreciate the improvement :)

Reply:
Fixed!